src/harness/externalCompileRunner.ts

```diff
@@ -0,0 +1,67 @@
+/// <reference path="harness.ts"/>
+/// <reference path="runnerbase.ts" />
+abstract class ExternalCompileRunnerBase extends RunnerBase {
+    abstract testDir: string;
+    public enumerateTestFiles() {
+        return Harness.IO.getDirectories(this.testDir);
+    }
+    /** Setup the runner's tests so that they are ready to be executed by the harness
+     *  The first test should be a describe/it block that sets up the harness's compiler instance appropriately
+     */
+    public initializeTests(): void {
+        // Read in and evaluate the test list
+        const testList = this.tests && this.tests.length ? this.tests : this.enumerateTestFiles();
+
+        describe(`${this.kind()} code samples`, () => {
+            for (const test of testList) {
+                this.runTest(test);
+            }
+        });
+    }
+    private runTest(directoryName: string) {
+        describe(directoryName, () => {
+            const cp = require("child_process");
+            const path = require("path");
+            const fs = require("fs");
+
+            it("should build successfully", () => {
+                const cwd = path.join(__dirname, "../../", this.testDir, directoryName);
+                const timeout = 600000; // 600s = 10 minutes
+                if (fs.existsSync(path.join(cwd, "package.json"))) {
+                    if (fs.existsSync(path.join(cwd, "package-lock.json"))) {
+                        fs.unlinkSync(path.join(cwd, "package-lock.json"));
+                    }
+                    const stdio = isWorker ? "pipe" : "inherit";
+                    const install = cp.spawnSync(`npm`, ["i"], { cwd, timeout, shell: true, stdio });
+                    if (install.status !== 0) throw new Error(`NPM Install for ${directoryName} failed!`);
+                }
+                Harness.Baseline.runBaseline(`${this.kind()}/${directoryName}.log`, () => {
+                    const result = cp.spawnSync(`node`, [path.join(__dirname, "tsc.js")], { cwd, timeout, shell: true });
+                    // tslint:disable-next-line:no-null-keyword
+                    return result.status === 0 ? null : `Exit Code: ${result.status}
+Standard output:
+${result.stdout.toString().replace(/\r\n/g, "\n")}
+
+
+Standard error:
+${result.stderr.toString().replace(/\r\n/g, "\n")}`;
+                });
+            });
+        });
+    }
+}
+
+class UserCodeRunner extends ExternalCompileRunnerBase {
+    public readonly testDir = "tests/cases/user/";
+    public kind(): TestRunnerKind {
+        return "user";
+    }
+}
+
+class DefinitelyTypedRunner extends ExternalCompileRunnerBase {
```

Comment:
This is starting to look like it should be just another instance of a class instead of its own class.

Reply:
True, except that Wesley is shortly going to want to put in more code into the UserCodeRunner, and I already need to put more code into the DefinitelyTypedRunner to skip ExpectErrors. I'll keep it the way it is for now and we can simplify later if neither of those efforts work out.